src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs

```diff
@@ -0,0 +1,274 @@
+        // /// <returns>A task that represents the asynchronous extraction operation.</returns>
+        // /// <remarks><para>Files of type <see cref="TarEntryType.BlockDevice"/>, <see cref="TarEntryType.CharacterDevice"/> or <see cref="TarEntryType.Fifo"/> can only be extracted in Unix platforms.</para>
+        // /// <para>Elevation is required to extract a <see cref="TarEntryType.BlockDevice"/> or <see cref="TarEntryType.CharacterDevice"/> to disk.</para></remarks>
+        // /// <exception cref="UnauthorizedAccessException">Operation not permitted due to insufficient permissions.</exception>
+        // public static Task ExtractToDirectoryAsync(string sourceFileName, string destinationDirectoryName, bool overwriteFiles, CancellationToken cancellationToken = default)
+        // {
+        //     throw new NotImplementedException();
+        // }
+
+        // Creates an archive from the contents of a directory.
+        // It assumes the sourceDirectoryName is a fully qualified path, and allows choosing if the archive stream should be left open or not.
+        private static void CreateFromDirectoryInternal(string sourceDirectoryName, Stream destination, bool includeBaseDirectory, bool leaveOpen)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(sourceDirectoryName));
+            Debug.Assert(destination != null);
+            Debug.Assert(Path.IsPathFullyQualified(sourceDirectoryName));
+            Debug.Assert(destination.CanWrite);
+
+            using (TarWriter writer = new TarWriter(destination, TarFormat.Pax, leaveOpen))
+            {
+                bool baseDirectoryIsEmpty = true;
+                DirectoryInfo di = new(sourceDirectoryName);
+                string basePath = di.FullName;
+
+                if (includeBaseDirectory && di.Parent != null)
+                {
+                    basePath = di.Parent.FullName;
+                }
+
+                // Windows' MaxPath (260) is used as an arbitrary default capacity, as it is likely
+                // to be greater than the length of typical entry names from the file system, even
+                // on non-Windows platforms. The capacity will be increased, if needed.
+                const int DefaultCapacity = 260;
+                char[] entryNameBuffer = ArrayPool<char>.Shared.Rent(DefaultCapacity);
+
+                try
+                {
+                    foreach (FileSystemInfo file in di.EnumerateFileSystemInfos("*", SearchOption.AllDirectories))
+                    {
+                        baseDirectoryIsEmpty = false;
+
+                        int entryNameLength = file.FullName.Length - basePath.Length;
+                        Debug.Assert(entryNameLength > 0);
+
+                        bool isDirectory = file.Attributes.HasFlag(FileAttributes.Directory);
+                        string entryName = ArchivingUtils.EntryFromPath(file.FullName, basePath.Length, entryNameLength, ref entryNameBuffer, appendPathSeparator: isDirectory);
```

Comment:
not related to your PR: I find it unintuitive that a buffer rented from array pool is passed by reference to a helper method. I needed to verify that the helper method returns the buffer to the pool in case it's too small.

Reply:
Is there anything I should do on my code to help improve its readability? Or maybe should we open an issue to discuss it separately?

Reply:
You can just do the rent/return inside EntryFromPath as an implementation detail.